pkg/fluxcsv/query_result.go

```diff
@@ -0,0 +1,297 @@
+	case "dateTime:RFC3339Nano":
+		return TimeDatatypeRFCNano, nil
+	default:
+		return InvalidDatatype, fmt.Errorf("unknown data type %s", s)
+	}
+}
+
+// QueryTableResult parses streamed flux query response into structures representing flux table parts
+// Walking though the result is done by repeatedly calling Next() until returns false.
+// Actual flux table info (columns with names, data types, etc) is returned by TableMetadata() method.
+// Data are acquired by Record() method.
+// Preliminary end can be caused by an error, so when Next() return false, check Err() for an error
+type QueryTableResult struct {
+	io.Closer
+	csvReader *csv.Reader
+
+	resultChanged      bool
+	tableIdChanged     bool
+	annotationsChanged bool
+
+	record *FluxRecord
+
+	baseColumns []*FluxColumn
+	metadata    *FluxTableMetadata
+	err         error
+}
+
+func NewQueryTableResult(rawResponse io.ReadCloser) *QueryTableResult {
+	csvReader := csv.NewReader(rawResponse)
+	csvReader.FieldsPerRecord = -1
+	return &QueryTableResult{Closer: rawResponse, csvReader: csvReader}
+}
+
+// ResultChanged returns true if the last call of Next() found a new query result
+func (q *QueryTableResult) ResultChanged() bool {
+	return q.resultChanged
+}
+
+// TableIdChanged returns true if the last call of Next() found a new table within the query result
+func (q *QueryTableResult) TableIdChanged() bool {
+	return q.tableIdChanged
+}
+
+// AnnotationsChanged returns true if last call of Next() found new CSV annotations
+func (q *QueryTableResult) AnnotationsChanged() bool {
+	return q.annotationsChanged
+}
+
+// Record returns last parsed flux table data row
+// Use Record methods to access value and row properties
+func (q *QueryTableResult) Record() *FluxRecord {
+	return q.record
+}
+
+// Metadata returns table-level info for last parsed flux table data row
+func (q *QueryTableResult) Metadata() *FluxTableMetadata {
+	return q.metadata
+}
+
+type parsingState int
+
+const (
+	parsingStateNormal parsingState = iota
+	parsingStateAnnotation
+	parsingStateNameRow
+	parsingStateError
+)
+
+// Next advances to next row in query result.
+// During the first time it is called, Next creates also table metadata
+// Actual parsed row is available through Record() function
+// Returns false in case of end or an error, otherwise true
+func (q *QueryTableResult) Next() bool {
+	var row []string
+	// set closing query in case of preliminary return
+	closer := func() {
+		if err := q.Close(); err != nil {
+			message := err.Error()
+			if q.err != nil {
+				message = fmt.Sprintf("%s,%s", message, q.err.Error())
+			}
+			q.err = errors.New(message)
+		}
+	}
+	defer func() {
+		closer()
+	}()
+	parsingState := parsingStateNormal
+	q.annotationsChanged = false
+	dataTypeAnnotationFound := false
+
+readRow:
+	row, q.err = q.csvReader.Read()
+	if q.err == io.EOF {
+		q.err = nil
+		return false
+	}
+	if q.err != nil {
+		return false
+	}
+
+	if len(row) <= 1 {
+		goto readRow
```

Comment:
Feels like I'd be remiss if I didn't make a comment about using `goto`...looks like it was brought over from `influxdata/influxdb-client-go` and it seems to work out fine here so I'm good with it. Might get ugly with a bunch of `break` and `continue`.

Reply:
I'll stare at it hard before merging to see how painful it'd be to remove.

Reply:
Hmmm, so I think it'd be possible to refactor into a `for{}` and use `continue` where all the `goto`s are now, but we'd still need to label the `for` so we could `break` it from inside of the `switch` statement. Not sure if it's worth adding the extra layer of indentation, I think `goto` is in Go to handle situations like this.